src/overlays/actors/ovl_Bg_Ydan_Hasi/z_bg_ydan_hasi.c

```diff
@@ -21,23 +27,180 @@ const ActorInit Bg_Ydan_Hasi_InitVars = {
+            this->actionFunc = BgYdanHasi_DecWaterTimer;
+        }
+        func_8002F948(&this->dyna.actor, NA_SE_EV_WATER_LEVEL_DOWN - SFX_FLAG);
+    }
+    waterBox = globalCtx->colCtx.stat.colHeader->waterBoxes;
+    globalCtx->colCtx.stat.colHeader->waterBoxes[1].unk_02 = this->dyna.actor.posRot.pos.y;
+    if (1) {}
+    return waterBox + 0x1;
+}
+
+void BgYdanHasi_DecWaterTimer(BgYdanHasi* this, GlobalContext* globalCtx) {
+    if (this->timer != 0) {
+        this->timer--;
+    }
+    func_8002F994(&this->dyna.actor, this->timer);
+    if (this->timer == 0) {
+        this->actionFunc = BgYdanHasi_MoveWater;
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Ydan_Hasi/func_808BE690.s")
+void BgYdanHasi_SetupThreeBlocks(BgYdanHasi* this, GlobalContext* globalCtx) {
+    if (Flags_GetSwitch(globalCtx, this->unk_168) != 0) {
+        this->timer = 260;
+        this->dyna.actor.draw = BgYdanHasi_Draw;
+        this->actionFunc = BgYdanHasi_UpdateThreeBlocks;
+        func_800800F8(globalCtx, 0xBE0, 0x1E, &this->dyna.actor, 0);
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Ydan_Hasi/func_808BE7C8.s")
+void BgYdanHasi_UpdateThreeBlocks(BgYdanHasi* this, GlobalContext* globalCtx) {
+    s32 pad;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Ydan_Hasi/func_808BE810.s")
+    if (this->timer != 0) {
+        this->timer--;
+    }
+    if (this->timer == 0) {
+        if (Math_ApproxF(&this->dyna.actor.posRot.pos.y, this->dyna.actor.initPosRot.pos.y, 3.0f) != 0) {
+            Flags_UnsetSwitch(globalCtx, this->unk_168);
+            this->dyna.actor.draw = NULL;
+            this->actionFunc = BgYdanHasi_SetupThreeBlocks;
+            return;
+        }
+        func_8002F948(&this->dyna.actor, NA_SE_EV_ELEVATOR_MOVE - SFX_FLAG);
+        return;
+    }
+    if (Math_ApproxF(&this->dyna.actor.posRot.pos.y, this->dyna.actor.initPosRot.pos.y + 120.0f, 3.0f) == 0) {
+        func_8002F948(&this->dyna.actor, NA_SE_EV_ELEVATOR_MOVE - SFX_FLAG);
+        return;
+    }
+    func_8002F994(&this->dyna.actor, this->timer);
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Ydan_Hasi/func_808BE8DC.s")
+void BgYdanHasi_Update(Actor* thisx, GlobalContext* globalCtx) {
+    BgYdanHasi* this = THIS;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Ydan_Hasi/func_808BE930.s")
+    this->actionFunc(this, globalCtx);
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Ydan_Hasi/func_808BE99C.s")
+void BgYdanHasi_Draw(Actor* thisx, GlobalContext* globalCtx) {
+    s32 pad;
+    GraphicsContext* gfxCtx;
+    Gfx* dispRefs[4];
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Ydan_Hasi/BgYdanHasi_Update.s")
+    if (thisx->params == 0 || thisx->params == 2) {
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Ydan_Hasi/BgYdanHasi_Draw.s")
+        Gfx_DrawDListOpa(globalCtx, D_808BEC24[thisx->params]);
```

Comment:
I'm pretty sure the pad is a THIS which should be used to access params here. We have thisx but I think we agreed to use this->actor.params for consistency.